fe/fe-core/src/main/java/com/starrocks/binlog/BinlogManager.java

```diff
@@ -0,0 +1,180 @@
+import org.apache.logging.log4j.Logger;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+public class BinlogManager {
+
+    private static final Logger LOG = LogManager.getLogger(BinlogManager.class);
+
+    SchemaChangeHandler schemaChangeHandler;
+
+    private Map<Long, List<Long>> binlogTxnIdToTableIds;
+
+    // When enable binlog, there may be concurrent imports
+    // record the map between all txnIds of concurrently imports
+    // and binlogTxnId
+    private Map<Long, Set<Long>> transcationidToBinlogTxnId;
+
+
+    private ReentrantReadWriteLock binlogLock = new ReentrantReadWriteLock(true);
+
+    public BinlogManager() {
+        binlogTxnIdToTableIds = new HashMap<>();
+        transcationidToBinlogTxnId = new HashMap<>();
+        schemaChangeHandler = GlobalStateMgr.getCurrentState().getSchemaChangeHandler();
+    }
+
+    public void addBinlogTxnIdToTableId(Long binlogTxnId, Long tableId) {
+        if (binlogTxnIdToTableIds.containsKey(binlogTxnId)) {
+            // should not happen, for the binlogTxnId can not be  duplicate
+            LOG.warn("the binlogTxnId has added to the binlog manager");
+            return;
+        }
+        binlogTxnIdToTableIds.putIfAbsent(binlogTxnId, new ArrayList<Long>()).add(tableId);
+    }
+
+    public boolean isBinlogWaitingTransactionId(Long transactionId) {
+        binlogLock.readLock().lock();
+        if (transcationidToBinlogTxnId.containsKey(transactionId)) {
+            return true;
+        }
+        binlogLock.readLock().unlock();
+        return false;
+    }
+
+    // the caller hold the write lock of Database
+    public void checkAndSetBinlogAvailableVersion(Long toPublishTxnId, Set<Long> runningTransactionId,
+                                                 Database db) {
+        binlogLock.readLock().lock();
+        for (Long binlogTxnId : transcationidToBinlogTxnId.get(toPublishTxnId)) {
+            boolean isLast = true;
+            for (Long transactionId : runningTransactionId) {
+                if (transactionId != toPublishTxnId && transactionId < binlogTxnId) {
+                    isLast = false;
+                }
+            }
+            if (isLast) {
+                List<Long> tableIds = binlogTxnIdToTableIds.get(binlogTxnId);
+                for (Long tableId : tableIds) {
+                    if (db.getTable(tableId).isOlapTable()) {
+                        OlapTable olapTable = (OlapTable) db.getTable(tableId);
+                        Collection<Partition> partitions =  olapTable.getPartitions();
+                        for (Partition partition : partitions) {
+                            // TODO persist the binlogAvailableVersion to ediglog
+                            partition.setBinlogAvailableVersion(partition.getVisibleVersion() + 1);
```

Comment:
Is this information need to be persisted ?

Reply:
If it‘s not persisted，the binlogAvaiableVersion will be set to current visibleVersion when master changes，so that the previously generated binlog is useless